core/src/main/java/com/linecorp/armeria/common/CompositeHttpHeadersBase.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.linecorp.armeria.common.StringMultimap.DEFAULT_SIZE_HINT;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.util.AsciiString;
+
+/**
+ * The composite container implementation which wraps {@link HttpHeadersBase}s
+ * to avoid expensive copy operations.
+ */
+class CompositeHttpHeadersBase
+        extends CompositeStringMultimap</* IN_NAME */ CharSequence, /* NAME */ AsciiString>
+        implements HttpHeaderGetters {
+
+    private static final Supplier<StringMultimap<CharSequence, AsciiString>> DEFAULT_APPENDER_SUPPLIER =
+            () -> new HttpHeadersBase(DEFAULT_SIZE_HINT);
+
+    CompositeHttpHeadersBase(HttpHeaderGetters... parents) {
+        super(from(parents), DEFAULT_APPENDER_SUPPLIER);
+    }
+
+    CompositeHttpHeadersBase(@Nullable CompositeStringMultimap<CharSequence, AsciiString> additionals,
+                             List<HttpHeaderGetters> parents,
+                             @Nullable CompositeStringMultimap<CharSequence, AsciiString> defaults) {
+        super(additionals, from(parents), defaults, DEFAULT_APPENDER_SUPPLIER);
+    }
+
+    private static List<StringMultimap<CharSequence, AsciiString>> from(HttpHeaderGetters... headers) {
+        if (headers.length == 0) {
+            return Collections.emptyList();
+        }
+
+        final ImmutableList.Builder<StringMultimap<CharSequence, AsciiString>> builder =
+                ImmutableList.builder();
+        for (HttpHeaderGetters header : headers) {
+            if (header instanceof HttpHeadersBase) {
+                builder.add((HttpHeadersBase) header);
+            } else {
+                builder.add(new HttpHeadersBase(header));
+            }
+        }
+
+        return builder.build();
+    }
+
+    private static List<StringMultimap<CharSequence, AsciiString>> from(List<HttpHeaderGetters> headers) {
+        if (headers.isEmpty()) {
+            return Collections.emptyList();
```

Comment:
Let's use immutableList for consistency.

Reply:
✅ updated! Replace all `Collections.emptyList()` to `ImmutableList.of()`.